File: 121.买卖股票的最佳时机/maxProfit.go
package problem0121

// 动态规划: 最大利润 = max{前一天的最大利润, 今天的价格-之前的最低价格}
func maxProfit(prices []int) int {
	if len(prices) <= 1 {
		return 0
	}
	min, max := prices[0], 0
	for i := 1; i < len(prices); i++ {
		if prices[i]-min > max {
			max = prices[i] - min
		}
		if prices[i] < min {
			min = prices[i]
		}
	}
	return max
}

File: 121.买卖股票的最佳时机/maxProfit.js
// 123的简化版本
function maxProfit(prices) {
  if (prices.length <= 1) {
    return 0
  }
  let buy1 = -prices[0]
  let sell1 = 0

  for (let i = 1; i < prices.length; i++) {
    buy1 = Math.max(buy1, -prices[i])
    sell1 = Math.max(sell1, prices[i] + buy1)
  }

  return sell1
}